nixos/modules/security/gnupg.nix

```diff
@@ -0,0 +1,270 @@
+            The path to the GnuPG-encrypted secret.
+            It will be copied into the Nix store of the orchestrating and of the target system.
+            It must be decipherable by an OpenPGP key within the GnuPG home <filename>/var/lib/gnupg/</filename>.
+            Defaults to the name of the secret,
+            prefixed by the path of the <link linkend="opt-security.gnupg.store">store</link>
+            and suffixed by <filename>.gpg</filename>.
+          '';
+        };
+        mode = lib.mkOption {
+          type = types.str;
+          default = "400";
+          description = ''
+            Permission mode of the secret <literal>path</literal>.
+          '';
+        };
+        user = lib.mkOption {
+          type = types.str;
+          default = "root";
+          description = ''
+            Owner of the secret <literal>path</literal>.
+          '';
+        };
+        group = lib.mkOption {
+          type = types.str;
+          default = "root";
+          description = ''
+            Group of the secret <literal>path</literal>.
+          '';
+        };
+        pipe = lib.mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          apply = x: if x == null then null else pkgs.writeShellScript "pipe" x;
+          description = ''
+            Shell script taking the deciphered secret on its standard input
+            and which must put on its standard output
+            the actual secret material to be installed.
+            This allows to decorate the secret with non-secret bits.
+          '';
+        };
+        path = lib.mkOption {
+          type = types.str;
+          default = name;
+          apply = p: if match "^/.*" p == null then "/run/keys/gnupg/"+p+"/file" else p;
+          description = ''
+            The path on the target system where the secret is installed to.
+            Default to the name of the secret,
+            prefixed by <filename>/run/keys/gnupg/</filename>
+            and suffixed by <filename>/file</filename>,
+            if non-absolute.
+          '';
+        };
+        service = lib.mkOption {
+          type = types.str;
+          default = "secret-" + escapeUnitName name + ".service";
+          description = ''
+            The name of the systemd service.
+            Useful to put constraints like <literal>after</literal> or <literal>wants</literal>
+            into services requiring this secret.
+          '';
+        };
+        postStart = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = "systemctl reload nginx.service";
+          description = ''
+            Commands to run after new secrets go live.
+            Typically the web server and other servers using secrets
+            need to be reloaded.
+          '';
+        };
+        postStop = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = ''shred -u "$secret_file"'';
```

Comment:
I think this may be a dangerous suggestion since people may uncritically copy this command to their own configuration without carefully considering whether it is effective on their file system.

Reply:
Thanks. I've moved the `postStart` and `postStop` options into `systemdConfig` which is also useful to set ` before`, `wantedBy` or `requiredBy` directly when configuring the secret instead of using `systemd.services.foo.after = [ secrets."path/to/secret".service ];`